### PR TITLE
fix: Disable leader election for webhooks

### LINKED ISSUE #2562: Karpenter webhook fails to acquire leader-election lease
### Version

__Karpenter Version:__ v0.16.2

__Kubernetes Version:__ v1.23.0

### Expected Behavior

Karpenter should startup without any issues and be able to be installed on clusters where the webhook leader election lease is left behind.

### Actual Behavior

Karpenter webhook crashes on startup and continues to crash as it fails to acquire the lease in time

### Steps to Reproduce the Problem

Install Karpenter in one namespace with 

```
helm install -n karpenter karpenter/karpenter --create-namespace
```

Attempt to move Karpenter to another namespace with

```
helm upgrade -n default karpenter/karpenter
```

### Resource Specs and Logs

```console
2022-09-27T21:52:12.180Z    INFO    webhook.ValidationWebhook    Starting controller and workers    {"commit": "346b7d7-dirty"}                                                                                    
2022-09-27T21:52:12.180Z    INFO    webhook.ValidationWebhook    Started workers    {"commit": "346b7d7-dirty"}                                                                                                    
2022-09-27T21:52:12.180Z    DEBUG    webhook.ValidationWebhook    Processing from queue validation.webhook.provisioners.karpenter.sh (depth: 1)    {"commit": "346b7d7-dirty"}                                     
2022-09-27T21:52:12.180Z    DEBUG    webhook.ValidationWebhook    Processing from queue karpenter/karpenter-cert (depth: 0)    {"commit": "346b7d7-dirty"}                                                         
I0927 21:52:12.180526       1 leaderelection.go:248] attempting to acquire leader lease karpenter/webhook.configmapwebhook.00-of-01...                                                                             
I0927 21:52:12.180764       1 leaderelection.go:248] attempting to acquire leader lease karpenter/webhook.webhookcertificates.00-of-01...                                                                          
I0927 21:52:12.180902       1 leaderelection.go:248] attempting to acquire leader lease karpenter/webhook.defaultingwebhook.00-of-01...                                                                            
I0927 21:52:12.181062       1 leaderelection.go:248] attempting to acquire leader lease karpenter/webhook.validationwebhook.00-of-01...                                                                            
2022/09/27 21:52:12 http: TLS handshake error from 169.254.175.250:46270: tls: no certificates configured                                                                                                          
2022-09-27T21:52:12.228Z    WARN    webhook    server key missing    {"commit": "346b7d7-dirty"}                                                                                                                   
2022-09-27T21:52:13.229Z    WARN    webhook    server key missing    {"commit": "346b7d7-dirty"}                                                                                                                   
2022/09/27 21:52:13 http: TLS handshake error from 169.254.175.250:46282: tls: no certificates configured                                                                                                          
2022/09/27 21:52:21 http: TLS handshake error from 169.254.175.250:39088: tls: no certificates configured                                                                                                          
2022-09-27T21:52:21.559Z    WARN    webhook    server key missing    {"commit": "346b7d7-dirty"}                                                                                                                   
2022-09-27T21:52:31.559Z    WARN    webhook    server key missing    {"commit": "346b7d7-dirty"}                                                                                                                   
2022/09/27 21:52:31 http: TLS handshake error from 169.254.175.250:53958: tls: no certificates configured                                                                                                          
2022/09/27 21:52:41 http: TLS handshake error from 169.254.175.250:36266: tls: no certificates configured 
```

__Relevant Slack Issues__
- https://kubernetes.slack.com/archives/C02SFFZSA2K/p1664041617803179
- https://kubernetes.slack.com/archives/C02SFFZSA2K/p1664315758307149

### Proposed Alternatives to Allow Leader Election

Quote from Slack:

_With existing leader record (lease object) present - not terribly unusual, especially in development - new clients have to wait the full LeaseDuration, [by design](https://github.com/kubernetes/client-go/blob/c81636cd4415f0858a99c2cc18023740df7aa15f/tools/leaderelection/leaderelection.go#L120-L129). This also means LeaseDuration should be as short as possible, core clients default to 15 seconds. [knative.dev/pkg/leaderelection](http://knative.dev/pkg/leaderelection) [defaults to 60 seconds](https://github.com/knative/pkg/blob/d1d5c849073b4c59fabeef1aaaf5eece021d9ee2/leaderelection/config.go#L124). This applies to all leader clients in Karpenter webhook process, including webhook certificates controller.
The problem is (in case of the existing leader records), while the webhook process waits for this leader election on startup, liveness and readiness probes on https-webhook are failing with "remote error: tls: unrecognized name", and liveness probe gives up before the lease duration expiration, so the webhook process is restarted before it has a chance to acquire the lease. (Karpenter controller is not affected - does not use https for probes, and controller-runtime uses lease of 15 seconds).
(Incidentally, I think the readiness probe part also explains error messages in controller that I was wondering about: "patching provisioner … failed calling webhook … no endpoints available for service karpenter")
Relaxing webhook liveness probe just a bit works (e.g. setting initialDelay to 60). A better fix might be to shorten the leases duration / renew deadline in webhook process (knative leaderelection supports passing config in context or via ConfigMap). In development loop, it may also make sense to cleanup stale leader records prior to redeployment with e.g. kubectl delete [leases.coordination.k8s.io](http://leases.coordination.k8s.io/) -n karpenter --all - this way the deployment stabilizes in seconds. (And providing a way to disable leader election in webhook process may still be useful for debugging?)_

_ Probably makes sense to not tie the liveness check (and potentially the webhook) to the leader election at all. as that would prevent any resilience to pod failure. Is the leader election process used for picking which karpenter pod is doing the scale up\down?_

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by jonathan-innis**
Fixes #2562 

**Description**

- Revert leader-election for webhooks with Karpenter in HA mode

**How was this change tested?**

* Manual install of Karpenter

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [x] No

**Release Note**

```release-note
None
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/webhooks/webhooks.go [hunk 1]
```diff
@@ -60,6 +60,7 @@ func Initialize(injectCloudProvider func(context.Context, cloudprovider.Options)
 		ServiceName: opts.KarpenterService,
 		SecretName:  fmt.Sprintf("%s-cert", opts.KarpenterService),
 	})
+	ctx = sharedmain.WithHADisabled(ctx) // Disable leader election for webhook
```
Comment (njtran) [inline pkg/webhooks/webhooks.go:63]: Can you either add a TODO here or adjust https://github.com/aws/karpenter/issues/2562 to reflect the need for more work after this?

**Review by njtran — APPROVED** [review]
LGTM! One nit for good measure

File: pkg/webhooks/webhooks.go (hunk 1 above)
Comment (jonathan-innis) [reply pkg/webhooks/webhooks.go:63]: I'll add a TODO, makes sense

Comment (jonathan-innis) [reply pkg/webhooks/webhooks.go:65]: Relevant details were added to the issue as well for the future

**Review by njtran — APPROVED** [review]
LGMT 🚢

**Pull request merged**